src/cpp/server/ServerMain.cpp

```diff
@@ -424,20 +432,51 @@ Error waitForSignals()
       else if (sig == SIGINT || sig == SIGQUIT || sig == SIGTERM)
       {
          LOG_DEBUG_MESSAGE("Received termination signal: " + std::to_string(sig));
-         //
-         // Here is where we can perform server cleanup e.g.
-         // closing pam sessions
-         //
+
+         Error error;
 
          // send SIGTERM signal to specific Workbench child processes
          // this way user processes will not receive the signal until it traverses the tree
-         std::set<std::string> interruptProcs =  {"rsession",
-                                                   "rserver-launcher",
-                                                   "rstudio-launcher",
-                                                   "rworkspaces",
-                                                   "rserver-monitor"};
-
-         Error error = core::system::sendSignalToSpecifiedChildProcesses(interruptProcs, SIGTERM);
+         std::set<std::string> interruptProcs = {
+            "rserver-launcher",
+            "rserver-monitor",
+            "rsession",
+            "rstudio-launcher",
+            "rworkspaces"
+         };
+         
+         // get list of child processes
+         std::vector<system::ProcessInfo> procInfos;
+         error = core::system::getChildProcesses(&procInfos, false);
+         if (error)
+         {
+            LOG_ERROR(error);
+            break;
+         }
+         
+         // pull out pids matching our process names
+         std::vector<pid_t> pids;
+         for (auto&& procInfo : procInfos)
+            if (interruptProcs.count(procInfo.exe))
+               pids.push_back(procInfo.pid);
+         
+         // create a thread that will wait for these processes
+         boost::thread waitThread([=]()
+         {
+            int status = 0;
+            for (pid_t pid : pids)
+            {
+               while (true)
+               {
+                  ::waitpid(pid, &status, 0);
+                  if (WIFEXITED(status))
+                     break;
+               }
+            }
+         });
```

Comment:
It looks like the rsession processes will still want to communicate with the rserver during suspend, so this PR lets rserver wait for those processes to clean up before exiting.

Reply:
I also wonder if we should have one thread per process, rather than a single thread waiting for all processes?

Reply:
I'd run this section of the PR past Jeff to get his thoughts on that # threads question, and just to be paranoid about any risks of stalling server shutdown, etc.